core/src/main/scala/tech/beshu/ror/accesscontrol/logging/AuditingTool.scala

```diff
@@ -55,21 +55,21 @@ class AuditingTool(settings: Settings,
     responseContext match {
       case allowedBy: ResponseContext.AllowedBy[B] =>
         AuditResponseContext.Allowed(
-          toAuditRequestContext(allowedBy.requestContext, Some(allowedBy.blockContext), allowedBy.history),
-          toAuditVerbosity(allowedBy.block.verbosity),
-          allowedBy.block.show
+          requestContext = toAuditRequestContext(allowedBy.requestContext, Some(allowedBy.blockContext), allowedBy.history),
+          verbosity = toAuditVerbosity(allowedBy.block.verbosity),
+          reason = allowedBy.block.show
         )
       case allow: ResponseContext.Allow[B] =>
         AuditResponseContext.Allowed(
-          toAuditRequestContext(allow.requestContext, None, allow.history),
-          toAuditVerbosity(Block.Verbosity.Info),
-          allow.block.show
+          requestContext = toAuditRequestContext(allow.requestContext, allow.history.collectFirst { case History(allow.block.name, _, blockContext) => blockContext }, allow.history),
```

Comment:
yes, great catch. But I'm not sure if it's safe to use block context from the matched block. Because  current_user_metadata is processed by ACL a little bit differently - processing is not finished after first matched block, but find all blocks which can be matched. The weird logic of further processing in `AccessControlList#handleMetadataRequest` method. When request is allowed, there is `UserMetadata` object created which is passed to `Allowed` object. So, we should use it instead of BlockContext here.

So, there three methods of `AuditRequestContextBasedOnAclResult` will be a little bit different in this case:

```
  override val loggedInUserName: Option[String] = blockContext.flatMap(_.userMetadata.loggedUser.map(_.id.value.value))
  override val impersonatedByUserName: Option[String] =
    blockContext
      .flatMap(_.userMetadata.loggedUser)
      .flatMap {
        case DirectlyLoggedUser(_) => None
        case ImpersonatedUser(_, impersonatedBy) => Some(impersonatedBy.value.value)
      }
  override val involvesIndices: Boolean = blockContext match {
    case Some(_: GeneralIndexRequestBlockContext | _: TemplateRequestBlockContext | _: SnapshotRequestBlockContext) => true
    case _ => false
  }
```

Reply:
I analyzed this flow, and I see, that difference between `block context's user metadata`, and `allowed's user metadata` is here tech.beshu.ror.accesscontrol.acl.AccessControlList#updateUserMetadataGroups. Difference is that, first user metadata is used in `RRMetadataResponse`, that is in rest response. Second user metadata is each, for block context. Serializing audit log event, we're in single rule context,  and it's history, not cumulative user metadata.

TLDR
There are tow user metadatas for two things in one place. That's misleading, but for me I'm using right one user metadata.

Reply:
yes, correct, but ... here in audit log, we see:
* request data
* history of ACL request processing
* data collected in matched block (block context)

We have also two types of request in ROR:
* regular ES request (travels through the blocks and stops when all rules in block match)
* current user metadata request (travels through all blocks, collects matched blocks and then gathers all user metadata found in matched block in one final metadata [different blocks can provides different metadata parts])

`AuditRequestContextBasedOnAclResult` was designed prior to the introduction of the current user metadata request handling described above (previously it was treated like regular ES request). All the logic how to create final user metadata object from fragmented metadata objects was encapsulated in `AccessControlList#handleMetadataRequest`. That's why I'm sure that here we should use the user metadata taked from `Allowed` object, not collecting first one from history.

Reply:
And please take a look at this:

```
override val involvesIndices: Boolean = blockContext match {
    case Some(_: GeneralIndexRequestBlockContext | _: TemplateRequestBlockContext | _: SnapshotRequestBlockContext) => true
    case _ => false
  }
```

It looks that it's a little bit outdates (and it's wrong atm). Maybe we should check here if given blockContext has `HasIndices` or `HasIndexPacks` type class instance.

Reply:
Done